File: benchmarks/cpu/mergesort/mergesort.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define N 200
#define ITERATIONS 50

long long next_rand(long long seed) {
    long long s = seed * 1103515245LL + 12345LL;
    if (s < 0) s = -s;
    s = s % 1000000007LL;
    return s;
}

void merge(long long *arr, int left, int mid, int right) {
    int n1 = mid - left;
    int n2 = right - mid;

    long long *L = (long long *)malloc(n1 * sizeof(long long));
    long long *R = (long long *)malloc(n2 * sizeof(long long));

    for (int i = 0; i < n1; i++)
        L[i] = arr[left + i];
    for (int i = 0; i < n2; i++)
        R[i] = arr[mid + i];

    int i = 0, j = 0, k = left;
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k++] = L[i++];
        } else {
            arr[k++] = R[j++];
        }
    }
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];

    free(L);
    free(R);
}

void mergesort(long long *arr, int left, int right) {
    if (right - left <= 1) return;
    int mid = left + (right - left) / 2;
    mergesort(arr, left, mid);
    mergesort(arr, mid, right);
    merge(arr, left, mid, right);
}

int main(void) {
    long long original[N];
    long long arr[N];

    /* Generate 200 pseudo-random values */
    long long seed = 42;
    for (int i = 0; i < N; i++) {
        seed = next_rand(seed);
        original[i] = seed % 10000;
    }

    /* Sort 50 times, copying the original array each time */
    for (int iter = 0; iter < ITERATIONS; iter++) {
        memcpy(arr, original, N * sizeof(long long));
        mergesort(arr, 0, N);
    }

    printf("%lld\n", arr[0]);
    return 0;
}

File: benchmarks/cpu/mergesort/mergesort.js
const N = 200;
const ITERATIONS = 50;

function nextRand(seed) {
    let s = seed * 1103515245 + 12345;
    if (s < 0) s = -s;
    s = s % 1000000007;
    return s;
}

function merge(left, right) {
    const result = [];
    let i = 0;
    let j = 0;
    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            result.push(left[i]);
            i++;
        } else {
            result.push(right[j]);
            j++;
        }
    }
    while (i < left.length) {
        result.push(left[i]);
        i++;
    }
    while (j < right.length) {
        result.push(right[j]);
        j++;
    }
    return result;
}

function mergesort(arr) {
    if (arr.length <= 1) return arr.slice();
    const mid = Math.floor(arr.length / 2);
    const left = mergesort(arr.slice(0, mid));
    const right = mergesort(arr.slice(mid));
    return merge(left, right);
}

// Use BigInt for the PRNG to avoid floating-point precision issues
function nextRandBig(seed) {
    let s = seed * 1103515245n + 12345n;
    if (s < 0n) s = -s;
    s = s % 1000000007n;
    return s;
}

function main() {
    const original = [];
    let seed = 42n;
    for (let i = 0; i < N; i++) {
        seed = nextRandBig(seed);
        original.push(Number(seed % 10000n));
    }

    let firstElement = 0;
    for (let iter = 0; iter < ITERATIONS; iter++) {
        const sorted = mergesort(original);
        firstElement = sorted[0];
    }

    console.log(firstElement);
}

main();
